ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -57,3 +91,180 @@ logTelemetry x = logPriority x Telemetry
 
 noLogging :: Logger
 noLogging = Logger $ \_ _ -> return ()
+
+data WithPriority a = WithPriority { priority :: Priority, payload :: a } deriving Functor
+
+-- | Note that this is logging actions _of the program_, not of the user.
+--   You shouldn't call warning/error if the user has caused an error, only
+--   if our code has gone wrong and is itself erroneous (e.g. we threw an exception).
+data Recorder msg = Recorder
+  { logger_ :: forall m. (HasCallStack, MonadIO m) => msg -> m () }
+
+logWith :: (HasCallStack, MonadIO m) => Recorder (WithPriority msg) -> Priority -> msg -> m ()
+logWith recorder priority msg = withFrozenCallStack $ logger_ recorder (WithPriority priority msg)
+
+instance Semigroup (Recorder msg) where
+  (<>) Recorder{ logger_ = logger_1 } Recorder{ logger_ = logger_2 } =
+    Recorder
+      { logger_ = \msg -> logger_1 msg >> logger_2 msg }
+
+instance Monoid (Recorder msg) where
+  mempty =
+    Recorder
+      { logger_ = \_ -> pure () }
+
+instance Contravariant Recorder where
+  contramap f Recorder{ logger_ } =
+    Recorder
+      { logger_ = logger_ . f }
+
+cmap :: (a -> b) -> Recorder b -> Recorder a
+cmap = contramap
+
+cmapWithPrio :: (a -> b) -> Recorder (WithPriority b) -> Recorder (WithPriority a)
+cmapWithPrio f = cmap (fmap f)
+
+cmapIO :: (a -> IO b) -> Recorder b -> Recorder a
```

Comment:
I'd probably call this `cmapM` (to match `co-log`), with the implicit constraint that we fix `m ~ IO` in our recorder. This is also fine though.

Reply:
I think in order to get `cmapM` working with the rank N type of `logger_` in `Recorder` I'd need an `UnliftIO` or `MonadBaseControl` constraint, and I remember having some issue there too.

I'm not sure if the rank N type of the `logger_` field is worth, but it does save carrying around a type parameter. Without the `logger_` having rank N type then using `co-log` directly would probably be better, although Pepe's initial example had a `tracer_` field bundled in which would be different from `co-log`.